akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/BehaviorSetup.scala

```diff
@@ -121,3 +121,59 @@ private[akka] object MDC {
     Map("persistenceId" -> persistenceId.id, "phase" -> phaseName)
   }
 }
+
+/**
+  * Setup snapshot and event delete/retention behavior. Retention bridges snapshot
+  * and journal behavior. This defines the retention criteria.
+  *
+  * @param snapshotEveryNEvents Snapshots are used to reduce playback/recovery times.
+  *                             This defines when a new snapshot is persisted.
+  *
+  * @param keepNSnapshots      After a snapshot is successfully completed,
+  *                             - if 2: retain last maximum 2 *`snapshot-size` events
+  *                             and 3 snapshots (2 old + latest snapshot)
+  *                             - if 0: all events with equal or lower sequence number
+  *                             will not be retained.
+  *
+  * @param deleteEventsOnSnapshot Opt-in ability to delete older events on successful
+  *                               save of snapshot. Defaults to disabled.
+  */
+/**
+  * Setup snapshot and event delete/retention behavior.
+  *
+  * INTERNAL API
+  */
+@InternalApi
+private[persistence]
+final case class RetentionCriteria private(snapshotEveryNEvents: Long,
+                                           keepNSnapshots: Long,
+                                           deleteEventsOnSnapshot: Boolean) {
+  /**
+    * Delete Messages:
+    *   {{{ toSequenceNr - keepNSnapshots * snapshotSize }}}
+    * Delete Snapshots:
+    *   {{{ (toSequenceNr - 1) - (keepNSnapshots * snapshotSize) }}}
+    *
+    * @param toSequenceNr the sequence number to delete to if `retainSnapshotsOnly` is false
+    */
+  def toSequenceNumber(toSequenceNr: Long): Long = {
+    // Delete old events, retain the latest
+    toSequenceNr - (keepNSnapshots * snapshotEveryNEvents)
+  }
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi
+private[persistence] object RetentionCriteria {
+
+  def apply(snapshotEveryNEvents: Long, keepNSnapshots: Long): RetentionCriteria =
```

Comment:
instead of `apply` I suggest that we name these factory methods, since we might add more for other algorithms later

```
val basedOnSnapshots: RetentionCriteria // the default

def basedOnSnapshots(keepNSnapshots:Int, deleteEventsOnSnapshot: Boolean): RetentionCriteria
```

Reply:
Shall we table that for my follow up PR for allowing multiple algorithms?